docs/coding.rst

```diff
@@ -312,79 +312,382 @@ The script will:
+
+In the above example we specify the available image sources using the ``srcset`` parameter. We then use ``sizes`` to say:
+
+- When the viewport is greater than ``1000px`` wide, the panda image will take up roughly half of the page width.
+- When the viewport is less than ``1000px`` wide, the panda image will take up roughly full page width.
+
+The browser will then pick the best source option from ``srcset`` (based on both the estimated image size and
+screen resolution) to satisfy the condition met in ``sizes``. Older browsers will fall back to the default image
+we specify using ``url``.
+
+Another example might be to serve a high resolution alternative for a fixed size image:
+
+.. code-block:: python
+
+    resp_img(
+        url="img/panda.png",
+        srcset={
+            "img/panda-high-res.png": "2x"
+        }
+    )
+
+This would output:
+
+.. code-block:: html
+
+    <img src="/media/img/panda.png" srcset="/media/img/panda-high-res.png 2x" alt="">
+
+Here we don't need a ``sizes`` attribute, since the panda image is fixed in size and small enough that it won't need to
+resize along with the browser window. Instead the ``srcset`` image includes an alternate high resolution source URL, along
+with a pixel density descriptor. This can then be used to say:
+
+- When a browser specifies a device pixel ratio of ``2x`` or greater, use ``panda-high-res.png``.
+- When a browser specifies a device pixel ration of less than ``2x``, use ``panda.png``.
+
+The ``resp_img()`` helper also supports localized images by setting the ``'l10n'`` parameter to ``True```:
+
+.. code-block:: python
+
+    resp_img(
+        url="img/panda-500.png",
+        srcset={
+            "img/panda-500.png": "500w",
+            "img/panda-750.png": "750w",
+            "img/panda-1000.png": "1000w"
+        },
+        sizes={
+            "(min-width: 1000px)": "calc(50vw - 200px)",
+            "default": "calc(100vw - 50px)"
+        },
+        optional_attributes={
+            "l10n": True
+        }
+    )
+
+This would output (assuming ``de`` was your locale):
+
+.. code-block:: html
+
+    <img src="/media/img/l10n/de/panda-500.png"
+         srcset="/media/img/l10n/de/panda-500.png 500w,/media/img/l10n/de/panda-750.png 750w,/media/img/l10n/de/panda-1000.png 1000w"
+         sizes="(min-width: 1000px) calc(50vw - 200px),calc(100vw - 50px)" alt="">'
+
+Finally, you can also specify any other additional attributes you might need using ``optional_attributes``:
+
+.. code-block:: python
+
+    resp_img(
+        url="img/panda-500.png",
+        srcset={
+            "img/panda-500.png": "500w",
+            "img/panda-750.png": "750w",
+            "img/panda-1000.png": "1000w"
+        },
+        sizes={
+            "(min-width: 1000px)": "calc(50vw - 200px)",
+            "default": "calc(100vw - 50px)"
+        },
+        optional_attributes={
+            "alt": "Red Panda",
+            "class": "panda-hero",
+            "height": 500",
```

Comment:
> should we make width and height required on images?

I'm open to making this enforced in the new helpers yeah, if others agree. Specifying both `height` and `width` attributes help to avoid layout shifts when pages are loading, or images lazy load. I'm not sure of any reasons why we might want to omit them (?), as doing so actively hurts page performance scores in tools like Lighthouse.

Thoughts @craigcook @reemhamz @nathan-barrett ?

Reply:
Yeah, I can't think of any good reason to omit them. +1 to enforcing good habits.

Reply:
I took a look at doing this but it may be more complex than it first seems, due to the way the `image()` macro is intertwined in all our other helpers. It might mean making `width` and `height `required attributes across the board, which I think is still a worthwhile change, but might likely blow this PR up in scope considerably.

It's probably best we file a followup issue for this 👍

Reply:
Followup sounds good